src/compiler/utilities.ts

```diff
@@ -5337,7 +5350,25 @@ namespace ts {
         const kind = node.kind;
         return isStatementKindButNotDeclarationKind(kind)
             || isDeclarationStatementKind(kind)
-            || kind === SyntaxKind.Block;
+            || isBlockStatement(node);
+    }
+
+    /* @internal */
+    export function isStatementOrBlock(node: Node): node is Statement {
+        const kind = node.kind;
+        return isStatementKindButNotDeclarationKind(kind)
+            || isDeclarationStatementKind(kind)
+            || node.kind === SyntaxKind.Block;
+    }
+
+    function isBlockStatement(node: Node): node is Block {

```

Comment:
I missed the `finally` case; your suggested implementation sounds much better.

The name is correct (or at least the best one I can come up with?). A *block statement* is a block that occurs in a grammatical context where a statement is allowed.